src/main/scala/io/opentargets/etl/backend/drug_beta/DrugExtensions.scala

```diff
@@ -58,6 +61,66 @@ object DrugExtensions extends LazyLogging {
 
   }
 
+  private def addCrossReferenceToMolecule(mols: DataFrame, references: DataFrame): DataFrame = {
+    logger.debug("Trying to add cross reference to molecule.")
+    // process references DF so that the xrefs are in a map.
+    val newXrefs: DataFrame = references
+      .withColumnRenamed("source", "key")
+      .groupBy("id", "key")
+      .agg(collect_list("reference").as("vNew"))
+
+    // add new xref column to molecule
+    val molXrefs = mols
+      .select(col("id"), explode(map_entries(col("crossReferences"))).as("m"))
+      .select(col("id"), col("m.key").as("key"), col("m.value").as("vOld"))
+
+    val joinedRefArrays = molXrefs
+      .join(newXrefs, Seq("id", "key"), "full_outer")
+      .withColumn("values",
+                  array_distinct(
+                    array_union(
+                      coalesce(col("vNew"), typedLit[Array[String]](Array.empty)),
+                      coalesce(col("vOld"), typedLit[Array[String]](Array.empty))
+                  ))
+      ).drop("vOld", "vNew")
+
+    val refMap = joinedRefArrays
+      .groupBy("id")
+      .agg(
+        map_from_arrays(
+          collect_list(col("key")),
+          collect_list(col("values"))
+        ).as("xrefs"))
+
+    mols
+      .join(refMap, Seq("id"), "full_outer")
+      .drop("crossReferences")
+      .withColumnRenamed("xrefs", "crossReferences")
+
+  }
+
+  /**
+    * processExtensions extracts a subset of extension files provided and prepares them for reading into dataframes.
+    *
+    * @param extentionType name of extension to filter for, eg synonym or cross-reference
+    * @param extensions list of all custom extension files provided in config
+    * @return filtered list of extensions in format ready to be read into dataframes.
+    */
+  private def groupExtensionByType(extentionType: String,
+                                   extensions: Seq[InputExtension]): Seq[IOResourceConfig] = {
+    extensions
```

Comment:
it could be simplified like
```scala
    val jsonPath = """(.+)\\.json$""".r

    extensions
      .filter(_.extensionType equalsIgnoreCase extentionType)
      .flatMap {
        ext =>
          ext.path match {
            case jsonPath(_*) => Some(IOResourceConfig("json", ext.path))
            case _ =>
              logger.error(s"Unable to process extension file ${ext.path.split("\\.").last}")
              None
          }
      }
```
or as I said before you can construct extractors for type of extensions and then do the same for the extension into a algebraic type like Anyextension -> JsonEx .... and use it in the extractor. extractors example in Capter 26 page 595 from Programming in Scala Third Edition. Nice example @cmalangone you can also have a look if you have time.

Reply:
the match is missing the optional `l`